devito/passes/clusters/aliases.py

```diff
@@ -81,244 +81,303 @@ def cire(clusters, mode, sregistry, options, platform):
-        ispace = cluster.ispace
-        for n in range(self._nrepeats(cluster)):
-            # Extract potentially aliasing expressions
-            mapper = self._extract(exprs, context, n)
+        assert len(exprs) == 0
 
-            # Search aliasing expressions
-            found = collect(mapper.extracted, ispace, self._opt_minstorage)
+        return processed
 
-            # Choose the aliasing expressions with a good flops/memory trade-off
-            exprs, chosen, pscore = choose(found, exprs, mapper, self._selector)
-            aliases.update(chosen)
-            score += pscore
+    def _do_generate(self, exprs, exclude, cbk_search, cbk_compose=None):
+        """
+        Carry out the bulk of the work of ``_generate``.
+        """
+        counter = generator()
+        make = lambda: Symbol(name='dummy%d' % counter())
 
-        # AliasMapper -> Schedule
-        schedule = lower_aliases(cluster, aliases, self._in_writeto, self._opt_maxpar)
+        if cbk_compose is None:
+            cbk_compose = lambda *args: None
 
-        # The actual score is a 2-tuple <flop-reduction-score, workin-set-score>
-        score = (score, len(aliases))
+        mapper = Uxmapper()
+        for e in exprs:
+            for i in cbk_search(e):
+                if not i.is_commutative:
+                    continue
 
-        return SpacePoint(schedule, exprs, score)
+                terms = cbk_compose(i)
 
-    def _make_symbol(self):
-        return Symbol(name=self.sregistry.make_name('dummy'))
+                # Make sure we won't break any data dependencies
+                if terms:
+                    free_symbols = set().union(*[i.free_symbols for i in terms])
+                else:
+                    free_symbols = i.free_symbols
+                if {a.function for a in free_symbols} & exclude:
+                    continue
+
+                mapper.add(i, make, terms)
+
+        return mapper
 
-    def _nrepeats(self, cluster):
+    def _generate(self, exprs, exclude):
+        """
+        Generate one or more extractions from ``exprs``. An extraction is a
+        set of CIRE candidates which may be turned into aliases. Two different
+        extractions may contain overlapping sub-expressions and, therefore,
+        should be processed and evaluated indipendently. An extraction won't
+        contain any of the symbols appearing in ``exclude``.
+        """
         raise NotImplementedError
 
-    def _extract(self, exprs, context, n):
+    def _lookup_key(self, c):
+        """
+        Create a key for the given Cluster. Clusters with same key may be
+        processed together in the search for CIRE candidates. Clusters should
+        have a different key if they must not be processed together, e.g.,
+        when this would lead to violation of data dependencies.
+        """
         raise NotImplementedError
 
-    def _in_writeto(self, dim, cluster):
+    def _betterv(self, delta_flops, delta_ws):
```

Comment:
name change?

Reply:
deliver typo

Reply:
agree, gonna find a better name, current one is leftover